File: frontend/weather/js/weather.js
const API_URL = 'http://localhost:5000/api';
let userCities = ['London', 'Tokyo', 'New York', 'Sydney'];
let currentCity = 'London';
let units = 'metric';

// Weather icons mapping
const weatherIcons = {
    '01d': '☀️', '01n': '🌙',
    '02d': '⛅', '02n': '🌥️',
    '03d': '☁️', '03n': '☁️',
    '04d': '☁️', '04n': '☁️',
    '09d': '🌧️', '09n': '🌧️',
    '10d': '🌦️', '10n': '🌧️',
    '11d': '⛈️', '11n': '⛈️',
    '13d': '❄️', '13n': '❄️',
    '50d': '🌫️', '50n': '🌫️'
};

// Initialize
document.addEventListener('DOMContentLoaded', () => {
    loadCurrentWeather(currentCity);
    loadForecast(currentCity);
    loadMultipleCities();
});

// Load current weather
async function loadCurrentWeather(city) {
    try {
        const response = await fetch(`${API_URL}/weather/current?city=${encodeURIComponent(city)}&units=${units}`);
        const data = await response.json();
        
        if (response.ok) {
            displayCurrentWeather(data);
            currentCity = city;
        } else {
            console.error('Error:', data.error);
        }
    } catch (error) {
        console.error('Error loading weather:', error);
    }
}

// Display current weather
function displayCurrentWeather(data) {
    const icon = weatherIcons[data.icon] || '🌤️';
    const tempUnit = units === 'metric' ? '°C' : '°F';
    const speedUnit = units === 'metric' ? 'm/s' : 'mph';
    
    const html = `
        <div class="weather-info">
            <div class="weather-main-info">
                <div class="weather-icon">${icon}</div>
                <div class="weather-details">
                    <div class="weather-city">${data.city}, ${data.country}</div>
                    <div class="weather-temp">${Math.round(data.temperature)}${tempUnit}</div>
                    <div class="weather-description">${data.description}</div>
                </div>
            </div>
            <div class="weather-stats">
                <div class="weather-stat">
                    <div class="stat-label">Feels Like</div>
                    <div class="stat-value">${Math.round(data.feels_like)}${tempUnit}</div>
                </div>
                <div class="weather-stat">
                    <div class="stat-label">Humidity</div>
                    <div class="stat-value">${data.humidity}%</div>
                </div>
                <div class="weather-stat">
                    <div class="stat-label">Wind Speed</div>
                    <div class="stat-value">${data.wind_speed} ${speedUnit}</div>
                </div>
                <div class="weather-stat">
                    <div class="stat-label">Pressure</div>
                    <div class="stat-value">${data.pressure} hPa</div>
                </div>
                <div class="weather-stat">
                    <div class="stat-label">Visibility</div>
                    <div class="stat-value">${(data.visibility / 1000).toFixed(1)} km</div>
                </div>
                <div class="weather-stat">
                    <div class="stat-label">Clouds</div>
                    <div class="stat-value">${data.clouds}%</div>
                </div>
            </div>
        </div>
    `;
    
    document.getElementById('current-weather').innerHTML = html;
}

// Load forecast
async function loadForecast(city) {
    try {
        const response = await fetch(`${API_URL}/weather/forecast?city=${encodeURIComponent(city)}&units=${units}`);
        const data = await response.json();
        
        if (response.ok) {
            displayForecast(data.forecast);
        }
    } catch (error) {
        console.error('Error loading forecast:', error);
    }
}

// Display forecast
function displayForecast(forecast) {
    const grid = document.getElementById('forecast-grid');
    grid.innerHTML = forecast.slice(0, 8).map(item => {
        const date = new Date(item.datetime);
        const icon = weatherIcons[item.icon] || '🌤️';
        const tempUnit = units === 'metric' ? '°C' : '°F';
        
        return `
            <div class="forecast-card">
                <div class="forecast-time">${date.toLocaleString('en-US', { month: 'short', day: 'numeric', hour: '2-digit', minute: '2-digit' })}</div>
                <div class="forecast-icon">${icon}</div>
                <div class="forecast-temp">${Math.round(item.temperature)}${tempUnit}</div>
                <div class="forecast-description">${item.description}</div>
                <div class="forecast-info">
                    💧 ${item.humidity}% | 💨 ${item.wind_speed} m/s
                </div>
            </div>
        `;
    }).join('');
}

// Load multiple cities
async function loadMultipleCities() {
    const grid = document.getElementById('cities-grid');
    grid.innerHTML = '';
    
    for (const city of userCities) {
        try {
            const response = await fetch(`${API_URL}/weather/current?city=${encodeURIComponent(city)}&units=${units}`);
            const data = await response.json();
            
            if (response.ok) {
                const icon = weatherIcons[data.icon] || '🌤️';
                const tempUnit = units === 'metric' ? '°C' : '°F';
                
                const card = document.createElement('div');
                card.className = 'city-card';
                card.innerHTML = `
                    <div class="city-card-header">
                        <div class="city-name">${data.city}</div>
                        <button class="remove-city" onclick="removeCity('${data.city}')">×</button>
                    </div>
                    <div class="city-icon">${icon}</div>
                    <div class="city-temp">${Math.round(data.temperature)}${tempUnit}</div>
                    <div class="city-weather">${data.description}</div>
                    <div class="city-stats">
                        <span>💧 ${data.humidity}%</span>
                        <span>💨 ${data.wind_speed} m/s</span>
                        <span>🧭 ${data.pressure} hPa</span>
                    </div>
                `;
                grid.appendChild(card);
            }
        } catch (error) {
            console.error(`Error loading weather for ${city}:`, error);
        }
    }
}

// Search weather
async function searchWeather() {
    const city = document.getElementById('city-search').value.trim();
    if (city) {
        loadCurrentWeather(city);
        loadForecast(city);
        document.getElementById('city-search').value = '';
        document.getElementById('search-suggestions').classList.remove('active');
    }
}

// Handle search keypress
function handleSearchKeypress(event) {
    if (event.key === 'Enter') {
        searchWeather();
    } else {
        searchCities(event.target.value);
    }
}

// Search cities with suggestions
async function searchCities(query) {
    if (!query || query.length < 2) {
        document.getElementById('search-suggestions').classList.remove('active');
        return;
    }
    
    try {
        const response = await fetch(`${API_URL}/weather/search?q=${encodeURIComponent(query)}&limit=5`);
        const cities = await response.json();
        
        if (response.ok && cities.length > 0) {
            const suggestions = document.getElementById('search-suggestions');
            suggestions.innerHTML = cities.map(city => `
                <div class="suggestion-item" onclick="selectCity('${city.name}')
                    title="${city.name}, ${city.country}">
                    <strong>${city.name}</strong>, ${city.country}
                    <br>
                    <small>${Math.round(city.temperature)}°C</small>
                </div>
            `).join('');
            suggestions.classList.add('active');
        }
    } catch (error) {
        console.error('Error searching cities:', error);
    }
}

// Select city from suggestions
function selectCity(city) {
    document.getElementById('city-search').value = city;
    searchWeather();
}

// Add city to favorites
function addCity() {
    const city = prompt('Enter city name:');
    if (city && !userCities.includes(city)) {
        userCities.push(city);
        loadMultipleCities();
    }
}

// Remove city from favorites
function removeCity(city) {
    userCities = userCities.filter(c => c !== city);
    loadMultipleCities();
}

// Toggle units
function toggleUnits() {
    units = units === 'metric' ? 'imperial' : 'metric';
    loadCurrentWeather(currentCity);
    loadForecast(currentCity);
    loadMultipleCities();
}

File: backend/app/routes/weather.py
"""Weather Routes"""

from flask import Blueprint, request, jsonify
import requests
from datetime import datetime
import os

weather_bp = Blueprint('weather', __name__, url_prefix='/api/weather')

# OpenWeatherMap API configuration
OPENWEATHER_API_KEY = os.environ.get('OPENWEATHER_API_KEY', 'YOUR_API_KEY_HERE')
OPENWEATHER_BASE_URL = 'https://api.openweathermap.org/data/2.5'

@weather_bp.route('/current', methods=['GET'])
def get_current_weather():
    """Get current weather for a location."""
    city = request.args.get('city', 'New York')
    units = request.args.get('units', 'metric')  # metric, imperial, standard
    
    if not OPENWEATHER_API_KEY or OPENWEATHER_API_KEY == 'YOUR_API_KEY_HERE':
        return jsonify({'error': 'Weather API key not configured'}), 500
    
    try:
        url = f'{OPENWEATHER_BASE_URL}/weather'
        params = {
            'q': city,
            'appid': OPENWEATHER_API_KEY,
            'units': units
        }
        
        response = requests.get(url, params=params, timeout=10)
        response.raise_for_status()
        
        data = response.json()
        
        return jsonify({
            'city': data['name'],
            'country': data['sys']['country'],
            'temperature': data['main']['temp'],
            'feels_like': data['main']['feels_like'],
            'humidity': data['main']['humidity'],
            'pressure': data['main']['pressure'],
            'wind_speed': data['wind']['speed'],
            'wind_direction': data['wind'].get('deg', 0),
            'clouds': data['clouds']['all'],
            'description': data['weather'][0]['main'],
            'icon': data['weather'][0]['icon'],
            'visibility': data.get('visibility', 0),
            'sunrise': datetime.fromtimestamp(data['sys']['sunrise']).isoformat(),
            'sunset': datetime.fromtimestamp(data['sys']['sunset']).isoformat(),
            'timezone': data['timezone'],
            'units': units
        }), 200
    
    except requests.exceptions.RequestException as e:
        return jsonify({'error': f'Failed to fetch weather: {str(e)}'}), 400

@weather_bp.route('/forecast', methods=['GET'])
def get_weather_forecast():
    """Get 5-day weather forecast for a location."""
    city = request.args.get('city', 'New York')
    units = request.args.get('units', 'metric')
    
    if not OPENWEATHER_API_KEY or OPENWEATHER_API_KEY == 'YOUR_API_KEY_HERE':
        return jsonify({'error': 'Weather API key not configured'}), 500
    
    try:
        url = f'{OPENWEATHER_BASE_URL}/forecast'
        params = {
            'q': city,
            'appid': OPENWEATHER_API_KEY,
            'units': units
        }
        
        response = requests.get(url, params=params, timeout=10)
        response.raise_for_status()
        
        data = response.json()
        
        forecast_list = []
        for item in data['list']:
            forecast_list.append({
                'datetime': item['dt_txt'],
                'temperature': item['main']['temp'],
                'feels_like': item['main']['feels_like'],
                'humidity': item['main']['humidity'],
                'pressure': item['main']['pressure'],
                'wind_speed': item['wind']['speed'],
                'description': item['weather'][0]['main'],
                'icon': item['weather'][0]['icon'],
                'clouds': item['clouds']['all']
            })
        
        return jsonify({
            'city': data['city']['name'],
            'country': data['city']['country'],
            'forecast': forecast_list,
            'units': units
        }), 200
    
    except requests.exceptions.RequestException as e:
        return jsonify({'error': f'Failed to fetch forecast: {str(e)}'}), 400

@weather_bp.route('/geo-weather', methods=['GET'])
def get_geo_weather():
    """Get weather by latitude and longitude."""
    lat = request.args.get('lat')
    lon = request.args.get('lon')
    units = request.args.get('units', 'metric')
    
    if not lat or not lon:
        return jsonify({'error': 'Latitude and longitude required'}), 400
    
    if not OPENWEATHER_API_KEY or OPENWEATHER_API_KEY == 'YOUR_API_KEY_HERE':
        return jsonify({'error': 'Weather API key not configured'}), 500
    
    try:
        url = f'{OPENWEATHER_BASE_URL}/weather'
        params = {
            'lat': lat,
            'lon': lon,
            'appid': OPENWEATHER_API_KEY,
            'units': units
        }
        
        response = requests.get(url, params=params, timeout=10)
        response.raise_for_status()
        
        data = response.json()
        
        return jsonify({
            'city': data['name'],
            'country': data['sys']['country'],
            'latitude': data['coord']['lat'],
            'longitude': data['coord']['lon'],
            'temperature': data['main']['temp'],
            'feels_like': data['main']['feels_like'],
            'humidity': data['main']['humidity'],
            'pressure': data['main']['pressure'],
            'wind_speed': data['wind']['speed'],
            'wind_direction': data['wind'].get('deg', 0),
            'clouds': data['clouds']['all'],
            'description': data['weather'][0]['main'],
            'icon': data['weather'][0]['icon'],
            'visibility': data.get('visibility', 0),
            'units': units
        }), 200
    
    except requests.exceptions.RequestException as e:
        return jsonify({'error': f'Failed to fetch weather: {str(e)}'}), 400

@weather_bp.route('/search', methods=['GET'])
def search_cities():
    """Search for cities."""
    query = request.args.get('q')
    limit = request.args.get('limit', 10, type=int)
    
    if not query:
        return jsonify({'error': 'Search query required'}), 400
    
    if not OPENWEATHER_API_KEY or OPENWEATHER_API_KEY == 'YOUR_API_KEY_HERE':
        return jsonify({'error': 'Weather API key not configured'}), 500
    
    try:
        url = f'{OPENWEATHER_BASE_URL}/find'
        params = {
            'q': query,
            'type': 'like',
            'appid': OPENWEATHER_API_KEY,
            'cnt': limit
        }
        
        response = requests.get(url, params=params, timeout=10)
        response.raise_for_status()
        
        data = response.json()
        
        cities = []
        for item in data.get('list', []):
            cities.append({
                'name': item['name'],
                'country': item['sys']['country'],
                'latitude': item['coord']['lat'],
                'longitude': item['coord']['lon'],
                'temperature': item['main']['temp']
            })
        
        return jsonify(cities), 200
    
    except requests.exceptions.RequestException as e:
        return jsonify({'error': f'Search failed: {str(e)}'}), 400
